src/index.ts

```diff
@@ -0,0 +1,116 @@
+
+// Connect an http socket to another tcp server.
+// Based on tunneling proxy code from https://nodejs.org/api/http.html
+export type NetConnectPromise = (port: number, hostname: string, cltSocket: net.Socket, req: Request) => Promise<net.Socket>;
+
+interface ConnectSocketOptions {
+	cltSocket: net.Socket;
+	hostname: string;
+	port: number;
+	head: Buffer;
+	connect: NetConnectPromise;
+	req: Request;
+}
+
+const connectSocket = ({ cltSocket, hostname, port, head, connect, req }: ConnectSocketOptions) =>
+	connect(port, hostname, cltSocket, req)
+	.then((srvSocket) => {
+		cltSocket.write('HTTP/1.0 200 Connection Established\r\nProxy-agent: Resin-VPN\r\n\r\n');
+		srvSocket.write(head);
+		srvSocket.pipe(cltSocket, {end: false});
+		cltSocket.pipe(srvSocket, {end: false});
+
+		return Promise.fromCallback((cb) => {
+			cltSocket.on('error', cb);
+			srvSocket.on('error', cb);
+			cltSocket.on('end', cb);
+			srvSocket.on('end', cb);
+		}).finally(() => {
+			srvSocket.destroy();
+			cltSocket.destroy();
+		});
+	}).tapCatch(() => {
+		cltSocket.end('HTTP/1.0 500 Internal Server Error\r\n');
+		cltSocket.destroy();
+	});
+
+// Create an http CONNECT tunneling proxy
+// Expressjs-like middleware can be used to change destination (by modifying req.url)
+// or for filtering requests (for example by terminating a socket early.)
+//
+// Returns an object with methods "listen" to start listening on a port,
+// and "use" for adding middleware.
+//
+// Middleware are functions of the form (request, controlSocket, head, next).
+export class Request extends http.IncomingMessage {
+	auth?: {
+		username?: string;
+		password?: string;
+	};
+}
+
+export class Tunnel extends EventEmitter {
+	private server = http.createServer((_req, res) => {
+		res.writeHead(405, {'Content-Type': 'text/plain'});
+		res.end('Method not allowed');
+	});
+	private middleware = new MiddlewareHandler() as MiddlewareHandlerAsync;
+
+	constructor() {
+		super();
+
+		this.server.on('connect', (req: Request, cltSocket: net.Socket, head: Buffer) =>
+			this.middleware.handleAsync([ req, cltSocket, head ])
+			.then(() => {
+				const srvUrl = url.parse(`http://${req.url}`);
+				return connectSocket({
+					cltSocket,
+					hostname: srvUrl.hostname!,
+					port: parseInt(srvUrl.port!, 10),
+					head,
+					connect: this.connect,
+					req,
+				})
+				.then(() =>
+					this.emit('connect', srvUrl.hostname, srvUrl.port, head));
+			}).catch((err) => {
+				this.emit('error', err);
+				cltSocket.destroy();
+			}));
+	}
+
+	connect(port: number, host: string, _cltSocket: net.Socket, _req: Request): Promise<net.Socket> {
+		const socket = net.connect(port, host);
+		return new Promise((resolve, reject) => {
+			socket.on('connect', () => resolve(socket));
+			socket.on('error', reject);
+		});
+	}
+	use = this.middleware.use.bind(this.middleware);
+	listen = this.server.listen.bind(this.server);
+	close = this.server.close.bind(this.server);
```

Comment:
This will result in `close` being any type. But `close` is actually of type: `(callback?: Function) => this`. Same goes for `listen` and `use` function accordingly.